### ISSUE TITLE
Signing on mobile v1

### ISSUE THREAD

**Issue by tschubotz** [gnosis/safe#348]
# What is this feature about? (1 sentence)
This will allow users to sign transactions of Safes they own with an existing seed phrase.

# Why is it needed? What is the value? For whom do we build it?
- We identified the problem area of not allowing users to sign easily on-the-go (cf. [doc).](https://docs.google.com/document/d/13EN-1B73sp3Gt3DfQTZ9zsCraX4clT6lFjL2kWcYRIQ/edit#)
- It's a strategic decision also: We believe that long term, mobile interaction will become more important.
- Signing with a mobile device is more convenient than with a hardware wallet. Also, people have their mobile phone with them anyways, no need to carry another device just for signing.
- This feature should provide a way for users to switch from their hardware wallet to the mobile device for signing.

# High-level overview of the feature
_Let's leave out generating keys/seedphrase including backing them up etc. out for now. Also let's leave out transaction executions. 
Instead, let's start with the bare minimum._

_Also, let's start with the import of 1 app-wide owner account in order to limit complexity. This leaves us the biggest flexibility later on, after we have learned how users want to use owner accounts on mobile._

### Import account for signing

- User can import an existing seed phrase to the app
  - 12 words AND 24 words should both work (Ledger and Trezor use 24 words, Metamask uses 12 words)
  - User can select which account to use/derive. (<- How complex is this?)
- Once successfully imported, the owner address is displayed somewhere in the app.
- For now, let's just allow the import of 1 account the can be used for all Safes, if it's an owner.
- User can import no matter if that account is an owner of any Safe or not.
- User can remove that account from the device again any time.

### Signing

- If the imported account is no owner, nothing changes.
- If the imported account is an owner, user can confirm or reject transactions that are waiting for confirmations.

# Tracking

- Please track the screen view events with `owner_enter_seed` and `owner_select_account` (cf. https://github.com/gnosis/safe/wiki/Overview-screen-tracking)

Please set the following new custom user property:

- num_keys_imported (Integer): Number of imported keys in the device. 0 if no key is imported, 1 if one key is imported.

# Misc

- Kick off doc: https://docs.google.com/document/d/1Q5Y3iicOnu9xwK05w5754JRhHxpHQxMyEI9RKj3ZWVQ/edit#
- "Press release": https://docs.google.com/document/d/1rzn4-GITyiai5bDt68PE2ya3RrJZmc9-_YZTDxKqXsM/edit#
- Tech requirements: https://docs.google.com/document/d/1pc-2aZyFCGRmwzwqexXd3I_YDmVTBcmGVai4HJQrGkE/edit#

# Updates
- Requirements for the sanitisation of the seed phrase:
  - Ignore whitespace and punctuation marks leading and following the input
  - Ignore extra whitespace (line breaks, tabs, spaces, etc) and punctuation symbols in the phrase. For example, words can be separated by more than one space or by new line symbol
  - Everything should be lower cased
  - Seed phrase validated according to underlying ETH/crypto library

Oct 2, Tobi: Added screen view tracking info.
Oct 6, Jose: Specs for seed phrase sanitising.
Oct 22, Dima: Updated tracking of the num_safe_owners after clarification

**Comment by tschubotz** [comment]
> Is it for you a must have feature to import multiple accounts at once for an MVP? (Even if not it probably makes sense to ask these questions already pre-emptively, but we might not have to answer all details about multi-account management on a technical and design level just yet)

I see the point that it's simpler to add just 1 app wide owner account.

But we specifically want to support multiple Safes, right? (No matter if owner or view only). That's something we learned a while ago. When the mobile app just had 1 Safe, we got the feedback that people wanted to have more than 1 Safe.

Besides that, it's not good for privacy if you reuse the same owner account in multiple Safes since there is an on-chain connection. We did learn during user research, that users are concerned about privacy. 

I feel it's wrong to make people reuse the same account. It definitely wrong, privacy-wise.

I'm really hesitant about this one, but I think we should really get in the mode of iterating and adding feature incrementally rather than adding too much right from the bet.

I propose the following changes to the epic in line with what Lukas suggested:

**(1) Add owner accounts via the app settings rather than the Safe specific settings since that is in line with how the web interface behaves, i.e. the web app has the wallet connection, not the Safe.

(2) Let's start with really just 1 account then since that would reduce the complexity. (i.e. we would not have to think about multiple accounts, which one to select etc.)**

@posthnikova @biafra23 @DmitryBespalov @MouazAlzahabi @rmeissner @lukasschor okay for you? If yes, I'll adapt the epic description.

**Comment by lukasschor** [comment]
What's the reason why you would do/access the key management via the settings?

I could also imagine at least two other options:

1)  Access it via the top-right corner of the header (same as on web) -> we would have to move the button to access the Safe list to the left, closer to the Safe address, for that. Which makes sense in my opinion anyways. 
2) Via a separate tab in the Navbar -> Not sure if I would like this though

Maybe they are also worth exploring?

**Comment by tschubotz** [comment]
> What's the reason why you would do/access the key management via the settings?

Doesn't have to be in the app settings, just wanted to say "not in the Safe settings". Happy to explore other places than app settings if it doesn't require reworking too much UI.

**Comment by posthnikova** [comment]
> (1) Add owner accounts via the app settings rather than the Safe specific settings since that is in line 
> with how the web interface behaves, i.e. the web app has the wallet connection, not the Safe.

> (2) Let's start with really just 1 account then since that would reduce the complexity. (i.e. we would not 
> have to think about multiple accounts, which one to select etc.)

@tschubotz For an MVP it's fine, we can introduce additional logic in the future (allow importing multiple accounts and activate one at a time app-wide or choose which account to use for which Safe). 

> I could also imagine at least two other options:

@lukasschor Floating action button could be an option.

**Comment by posthnikova** [comment]
These are all options to access owner accounts:
1) From app settings
2) From settings (3 tabs instead of 2)
3) From the Nav bar (2 buttons)
4) From the Nav bar - bottom sheet has 2 tabs
5) Floating action button
6) From tab bar (4 tabs instead of 3)

![image](https://user-images.githubusercontent.com/45787536/91035518-afb09900-e606-11ea-9c5f-09059fc52aea.png)

**Comment by lukasschor** [comment]
Nice! Some initial thoughts I had.

App Settings
- Seems like this requires the least changes & assumptions in terms of the UX
- The feature is a bit more stealth this way, which is good in case we are not yet fully sure about the importance / future implementation
- Discoverability is very low, it looks "buried away"

Third Settings
- Looks unclear to me whether the owners that are added are Safe-specific or app-wide, which can also be a good thing as users would then complain in case it does not behave as they would have expected.

Nav Bar 2 Buttons
- The Safelist button seems too close to the owner identicon
- We learned on Web that 2 Identicons are confusing, maybe just show a "key" or "silhouette" icon
- Maybe the Safe Swtich button could be moved into the Receive screen

Nav Bar Tabbed Bottom Sheet
- I like that all account "switching" is done in the same place
- Here the distinction between Safes and keys might be less clear

Floating button
- Would not show a Wallet icon (we don't have this info anyways) but instead a "key" or "silhouette" icon
- I'm generally not a big fan of floating buttons but that's a very personal view
- Floating buttons should be for a very common / recurring action and I don't think switching owners is quite that important

Tab Bar
- I would always have Settings as the option at the very right of the bar
- Might be too prominent of a placement for this feature, might make sense in case key management becomes a bigger part of the app (generating keys, deriving different accounts from the same seed, etc.)

**Comment by tschubotz** [comment]
Super nice that you made so many different versions!

"App settings" , "Settings 3 Tabs"

- I agree with Lukas' points, nothing to add from my end.

"Nav Bar 2 Buttons"

- This would copy the desktop-Metamask way of integration. 
- I feel it's not necessary to display an indication all the time, which owner account is currently active since most functionality is read-only.
- This crowds the nav bar.

"Nav Bar Tabbed Bottom Sheet"

- Makes it clear that the owner account selection is independent of the Safe account selection.
- Provides for enough space for potential future multiple accounts.

"Floating action button"

- There's probably no way to make the feature more prominent.
- I would not do that for this feature.

"Tab bar"

- Puts the owner on the same level like Settings, assets and txs. The latter 2 are Safe specific. The settings not. Not ideal to add owners there imo.

My conclusion

- I like "Settings" and "bottom sheet" the most. 
- I prefer bottom sheet a bit since it makes it clear that owners are something in parallel to Safes and it's less buried away.

**Comment by posthnikova** [comment]
**Import**

![image](https://user-images.githubusercontent.com/45787536/91707159-b8ade700-eb7f-11ea-86e5-ce0f05bdccd2.png)

**Entry point**
For the first version I prefer 'Tabbed bottom sheet'. It's prominent enough and clear that the wallet is connected app-wide. It's hard to say what the implications for future versions would be. 
**Read-only**
I chose to indicate which Safes are read-only rather than Safes which are non-read-only because it's aligned with the web interface. 
**Account or wallet?**
I decided to call it 'owner wallet' instead of 'account', 'owner account' or 'key' because in the web interface we call it 'wallet', 'connect wallet'. 

**Sign**

![image](https://user-images.githubusercontent.com/45787536/91707308-e4c96800-eb7f-11ea-8e43-f0ff7857ff88.png)

**Remove**

![image](https://user-images.githubusercontent.com/45787536/91707514-3245d500-eb80-11ea-8b7f-dd15fc2b33cb.png)

**Some open questions**
1) Do we need to enter custom name for an owner wallet? Edit name?
2) Will there be biometrics when signing tx? 
3) Do we have to choose between 12 or 24 word seed phrase or it's detected automatically? 
4) Can we detect if it's a hardware wallet or Metamask?
5) Can we show which owner is you in owner addresses in Safe settings? https://prnt.sc/u8uuwr
6) Do we allow importing an owner before adding any Safes? 
7) What happens when all Safes are removed but the owner is not removed?
8) What if the owners change? How often do we check it? Shall we get notification when the imported wallet is no longer an owner?
9) Can we show in tx details who of the owners is you? https://prnt.sc/u8uwye

**Comment by sche** [comment]
For me, it is a bit cumbersome to add/remove an owner from a sheet of switching safes. It is not obvious, and the use cases are too different to put them close to each other.

I would put the owner’s wallet in a “Switch Safe” screen only if we would treat it similar to a Safe (e.g. user could select it to view balances, transactions, etc.)

But as this is not our plan, I would be for having it in App settings. Though I don’t like much that this way it is a bit hidden. Probably we could add the “Add owner” button to the Safe Info screen, that would be open by tapping on “Read-only” in the header view anyways.

> Will there be biometrics when signing tx?
We decided to leave it out for later.

> Do we have to choose between 12 or 24 word seed phrase or it's detected automatically?
In the legacy app, we had 12 words phrases only. I would leave it like that.

> Can we detect if it's a hardware wallet or Metamask?
No

> Do we allow importing an owner before adding any Safes?
> What happens when all Safes are removed but the owner is not removed?
I would allow importing the owner’s wallet separately.

> Shall we get notification when the imported wallet is no longer an owner?
The backend would need to know the app’s owner.

**Comment by rmeissner** [comment]
> In the legacy app, we had 12 words phrases only. I would leave it like that.

We decided that both need to be able to input as ledger for example uses 24 words as their seed phrase. Also normally the words are lowercased/ normalized and then you just hash them, so it doesn't matter if it is 3, 6, 9, 12, 15, 18,,21 or 24 (but it needs to be a multiple of 3 apparently :D). The android lib should be able to handle them all.

> Can we detect if it's a hardware wallet or Metamask?

All private key based accounts are essentially the same, so there is no way to distinguish them.

> Shall we get notification when the imported wallet is no longer an owner?

Notifications are handled in a different issue, as we need to solve authentication for that.

**Comment by tschubotz** [comment]
Thanks for the wireframes @posthnikova 

I think it makes sense to have the owner wallet management close to the Safe management in the bottom sheet.

> Do we need to enter custom name for an owner wallet? Edit name?

I'd say this is a nice to have, not a must have, would leave out in this version

> Will there be biometrics when signing tx?

I currently agree to Andrey: Not in this version. 
There was a pending open question whether on Android we can use the system biometric input easily to confirm messages though.

> Do we have to choose between 12 or 24 word seed phrase or it's detected automatically?

We could build it either way. Support for 12 and 24 words is the minimum requirement since MM uses 12 words and ledger uses 24 words.

> Can we detect if it's a hardware wallet or Metamask?

Unfortunately not.

> Can we show which owner is you in owner addresses in Safe settings? https://prnt.sc/u8uuwr

That should be possible, from technical perspective.

> Do we allow importing an owner before adding any Safes?

Good question. Would both work. I tend to: yes, since then the owner wallet could be retained when removing all Safes from the device. But I would go with what easier here.

> What happens when all Safes are removed but the owner is not removed?

I think the owner wallet could be retained in that case.

> What if the owners change? How often do we check it? Shall we get notification when the imported wallet is no longer an owner?

I think the app should handle that silently, i.e. no notification necessary etc.

> Can we show in tx details who of the owners is you? https://prnt.sc/u8uwye

I think so, yes.

**Comment by tschubotz** [comment]
![image](https://user-images.githubusercontent.com/1380529/91734048-b1500300-ebaa-11ea-9b45-f8925f8fcda3.png)

I like what Trust wallet writes as description: "Typically 12 (sometimes 24) words separated by single spaces"

**Comment by posthnikova** [comment]
New version. 

**Import (from bottom sheet)**
What's been changed:
• Added a note that the phrase could be 12 or 24 words @tschubotz
• Added text 'Now you can confirm transactions for all Safes controlled by this owner wallet (without execution)' on success screen. 
• Show which owner is imported among owner addresses (in Safe settings)

![image](https://user-images.githubusercontent.com/45787536/91990614-3ca7d080-ed32-11ea-9878-76a429eaff13.png)

**Sign**
What's been changed:
• Owner shows in confirmation timeline as 'Imported wallet'

![image](https://user-images.githubusercontent.com/45787536/91990636-43cede80-ed32-11ea-8556-d1a7cb036741.png)

**Remove**
Nothing changed

**Import wallet when there are no Safes loaded**

![image](https://user-images.githubusercontent.com/45787536/91990960-a6c07580-ed32-11ea-9949-c92973220d6b.png)

**Import - a version from app settings**

![image](https://user-images.githubusercontent.com/45787536/91991372-1a628280-ed33-11ea-848a-cb3f287b0fc4.png)

@sche This is less preferred because this option is hidden. Maybe we suggest to import wallet right after the user loaded the first Safe? Or in case there are Safes show a banner on Balances screen to improve discoverability?

![Uploading image.png…]()

**Comment by tschubotz** [comment]
> **Import wallet when there are no Safes loaded**

I think that's a nice solution.

![image](https://user-images.githubusercontent.com/1380529/91992378-5b0ecb80-ed34-11ea-99cf-2b3e9d104292.png)

I think just "hardware" sounds bit weird currently, perhaps something got lost. How about the following:
"Import the seed phrase e.g. from your hardware wallet or Metamask owner. Typically 12 (sometimes 24) words separated by single spaces."

![image](https://user-images.githubusercontent.com/1380529/91992795-d96b6d80-ed34-11ea-9be8-18f34cf4f7ef.png)

Nice ideas. Depending on the implementation effort I would only go with none or only 1 of them though, for v1.

**Comment by sche** [comment]
@posthnikova I like the Banner idea.
For the option with adding a wallet from app settings, I think it makes sense to add a call to action "Import Wallet" button on the safe info screen.

**Comment by lukasschor** [comment]
Not sure if we should say "Import Wallet" but rather "Import Owner"? Because you're not really importing a wallet. See here for more background on this: https://twitter.com/pedrouid/status/1295789310042673157?s=20

Opening bit of a pandora's box here:
In general, the language around this is not very well streamlined. Even on the web interface we sometimes call it "owner", "wallet", "owner wallet", "connected wallet" etc. This has grown organically and we don't really have some insights to justify a specific wording vs. another. For example, why do we not call it "signer", "key" or "signing key"  etc., which would also make more sense than "owner" in a single-user use-case and also go better with the "Safe" analogy? I don't think we should have this discussion necessarily with this feature. But thought it would be an opportunity to raise the issue for future considerations.

**Comment by KristinaMayman** [comment]
I also like the idea of alerting the user about the possibility to import a wallet, making that feature more visible at least at first.

As for the terminology, I haven't noticed a strong preference among the users. So far in the tests they have followed the language used by the interface. We had a couple of mobile prototype tests that used the word "key" specifically and then the users followed suit. But in almost all of the web app tests, "owner" or "wallet" was used.

**Comment by posthnikova** [comment]
The latest version after refinement meeting is on Invision. 

iOS: https://invis.io/4EYSS957HQX
Android: https://invis.io/A3YSSA5TQ9E

What's been changed:
• Added popup 'Confirm transaction' to protect from accidental confirmation
• Show 20 derived accounts

**Comment by posthnikova** [comment]
Changes after user tests. The recommendations were:
**• Rename 'contract version' → 'Safe version' in the Safe settings**

![image](https://user-images.githubusercontent.com/45787536/95728844-4d762b00-0c7c-11eb-8425-13c3f49a9dcc.png)

**• Hide derived keys under 'Show more keys'** 

![image](https://user-images.githubusercontent.com/45787536/95728877-5830c000-0c7c-11eb-8554-17049a04a0a1.png)

**• Move entry point to Safe settings**
I decided not to do this because non-owner key could show up for some Safes https://prnt.sc/urty7c

**• Add info about where the execution needs to happen** 
I have 2 versions. The first option is to add info screen with steps and where the user is in the process (similar to was proposed for the web interface https://invis.io/T4XF6KSQX7M#/428161605_Transactions_-_Approve_Popup_1272px_). There is info that the execution is not possible in this app.

![image](https://user-images.githubusercontent.com/45787536/95729019-857d6e00-0c7c-11eb-81f3-a70584a00b16.png)

The second option is to divide this process into 2 steps. Info about the execution is on a separate screen. 

![image](https://user-images.githubusercontent.com/45787536/95729052-8f9f6c80-0c7c-11eb-87da-5ffcddb89b5b.png)

**• Add info that importing a key will not add anything to the assets**
I though about where we can show this info and the best option I see is to show 'onboarding' slides (although I'm not a big fan of them). Other options were to add this info to Enter Seed screen, or to show tooltip or banner after importing owner. 

![image](https://user-images.githubusercontent.com/45787536/95729256-ed33b900-0c7c-11eb-98b5-a34a75c872c1.png)

**Comment by sche** [comment]
thank you @posthnikova 
I like more explanatory screens. I like the 2nd confirmation option more. My only comment there would be is that the networking (request submission) should happen before the "Got it" screen displayed.

**Comment by KristinaMayman** [comment]
Thanks for posting these! I think hiding the derived keys looks very good. 

I also have a slight preference for the second version of the no-execution flow, since it might be easy to just click through the first version without noticing the message about the execution (even though the "under construction" symbol should attract some attention). My only concern is that there is still no information for the user about the execution once the confirmation is done. Could we maybe add "should be done from the web interface" or something like that under the "Execute" in the timeline on the last screen?

The explanatory screens might be helpful for some... We should definitely add an option to skip them though.

**Comment by jpalvarezl** [comment]
> • Move entry point to Safe settings
> I decided not to do this because non-owner key could show up for some Safes https://prnt.sc/urty7c

Good decision. AFAICT, a user could potentially setup the same owner key for a few safes, in which case this should definitely be regarded as a setting of "the app" rather than "safe-specfic".

> • Add info that importing a key will not add anything to the assets
> I though about where we can show this info and the best option I see is to show 'onboarding' slides (although I'm not a big fan of them). Other options were to add this info to Enter Seed screen, or to show tooltip or banner after importing owner.

Maybe in the last slide we could have something like a checkbox saying "Don't show me this information next time" or something. In case a user changes a bit too often keys and gets tired of going through that flow. 

> • Add info about where the execution needs to happen
Option seems more concise and scalable. If eventually we can confirm and execute from the app, the changes would be minimal in terms of UI. 

Good work!

**Comment by posthnikova** [comment]
New iteration after comments. What's been done:

**Import owner**
I added an option to skip onboarding slides. The slides should only be shown once to avoid embarrasment. 

![image](https://user-images.githubusercontent.com/45787536/96705196-41cde700-1395-11eb-9e16-e85efc76602f.png)

As an alternative, we can show this info on one screen before entering seed:

![image](https://user-images.githubusercontent.com/45787536/96705320-688c1d80-1395-11eb-8e6d-18c787241819.png)

**Sign transaction**
Most people chose the second option (divide the process into 2 steps, info about the execution is on a separate screen). I think we should only show the 'execute' step when all confirmations are collected (this is the last confirmation):

![image](https://user-images.githubusercontent.com/45787536/96706016-51016480-1396-11eb-9f73-f3e62e7e692e.png)

When this is not the last confirmation the logic would be different. Only 'collect confirmations' step is shown:

![image](https://user-images.githubusercontent.com/45787536/96706171-87d77a80-1396-11eb-843b-0f05396a00e8.png)

Info that execution is not available in the app should be shown when all confirmations are collected. 

> My only comment there would be is that the networking (request submission) should happen before 
> the "Got it" screen displayed.

@sche So request happens after we hit 'Confirm transaction'. Where does the loading happen then? Does that mean there will be loading state before we can show 'execute' step?

> Could we maybe add "should be done from the web interface" or something like that under 
> the "Execute" in the timeline on the last screen?

@KristinaMayman I added this. 

> We should definitely add an option to skip them though.

There will be 'Skip' button.

> Maybe in the last slide we could have something like a checkbox saying "Don't show me this 
> information next time" or something. In case a user changes a bit too often keys and gets tired 
> of going through that flow.

@jpalvarezl I think we should only show this once, plus you can skip the slides.

**Comment by rmeissner** [comment]
Thanks for the updates @posthnikova looks very nice.

Regarding the onboarding screen:
There is a typo "we do not store you**r** private key". Also we actually store only the private key and not the seed phrase, as the seed phrase makes it possible to generate the private key.

Also did we check which hardware wallets we support? Because last time we checked we were no sure that we support all trezor and ledger variants.

Personal opinion for the confirmation screen (with the steps): I was confused why the steps are greyed out in the beginning. Also not sure about the difference of "Execute Transaction" and "Submit to chain"

**Comment by DmitryBespalov** [comment]
All proposed changes look good to me.

**Comment by KristinaMayman** [comment]
I added screen-specific comments on Invision but I actually like having onboarding on cards. Maybe that would make it less likely for the user to mindlessly click through if there is an option to scan through the three headlines and pick out the relevant information.

**Comment by liliya-soroka** [comment]
Verified
app version
ios -2.5.0 (186)
android - 2.5.0 -1421rc
